### ISSUE TITLE
Problems with contacting 3rd party API from Parse Nodejs backend

### ISSUE THREAD

**Issue by nharbo** [parse-community/parse-server#4336]
### Issue Description

I'm using Parse as a backend for my web-app. It's hosted on Amazon.
I have to send an email from my node-js server, through Amazons SNS, which i'm doing with the npm-package 'node-ses'.

I set up a cloudcode function, taking my parameters as informations for the content. 

This is how my cloudcode function looks:

```
Parse.Cloud.define('buy-giftcard', function (request, response) {

  return response.success("Hello!");

  client.sendEmail({
    to: request.params.buyerEmail
    , from: 'hello@xxxx.xxx'
    , cc: ""
    , bcc: ""
    , subject: 'Congratulations'
    , message:  request.params.message
    , altText: 'plain text'
   }, function (err, data, res) {
    console.log(err);
    console.log(data);
    console.log(res);
   });

});
```

#### Expected Results

When the cloudcode is called from my react-app (also hosted on AWS) I expect the email to be sent, and no response, as I don't add any in the cloudcode function. 

#### Actual Outcome

The email is sent 5 times, and the "No 'Access-Control-Allow-Origin' header is present on the requested resource" error shows 5 times - I guess it's the limit for retries or something like that. 

I get this response in my developerconsole:

Failed to load https://dev-api.xxxxxx.xxx/parse/functions/buy-giftcard: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:3000' is therefore not allowed access. The response had HTTP status code 504.

This is the error I get in my Parse server-log:

[error] 5414#0: *20 connect() failed (111: Connection refused) while connecting to upstream, client: 172.31.12.25, server: , request: "POST /parse/functions/buy-giftcard HTTP/1.1", upstream: "http://127.0.0.1:8081/parse/functions/buy-giftcard", host: "dev-api.xxx.xxxx", referrer: "http://localhost:3000/store"

If I remove the client.sendEmail from the function, and just returns something like response.success("Hello"); everything works as expected. So I guess it's something about the SNS call that makes it go wrong.

Does anyone have an idea about what this could be?

### Environment Setup

- **Server**
  - parse-server version : 1.9.2
  - Hosted with AWS

- **Database**
  - MongoDB version: don't know, but I guess it's not important for this issue :)
  - Hosted with mLab

### Logs/Trace

2017/11/01 10:02:12 [warn] 22960#0: duplicate MIME type "text/html" in /etc/nginx/conf.d/00_elastic_beanstalk_proxy.conf:42
2017/11/01 10:37:25 [warn] 25031#0: duplicate MIME type "text/html" in /etc/nginx/conf.d/00_elastic_beanstalk_proxy.conf:42
2017/11/01 14:10:47 [warn] 302#0: duplicate MIME type "text/html" in /etc/nginx/conf.d/00_elastic_beanstalk_proxy.conf:42
2017/11/01 14:41:40 [warn] 2306#0: duplicate MIME type "text/html" in /etc/nginx/conf.d/00_elastic_beanstalk_proxy.conf:42
2017/11/04 01:27:01 [error] 2317#0: *85886 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 172.31.12.25, server: , request: "POST /parse/classes/UserData/N8KXYtGFrf HTTP/1.1", upstream: "http://127.0.0.1:8081/parse/classes/UserData/N8KXYtGFrf", host: "dev-api.xxxx.xxxx"
2017/11/04 01:40:01 [error] 2317#0: *86504 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 172.31.29.236, server: , request: "POST /parse/classes/UserData/NgrgxOngKD HTTP/1.1", upstream: "http://127.0.0.1:8081/parse/classes/UserData/NgrgxOngKD", host: "dev-api.xxxx.xxxxx"
2017/11/04 12:07:01 [error] 2317#0: *116659 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 172.31.29.236, server: , request: "POST /parse/classes/UserData/NgrgxOngKD HTTP/1.1", upstream: "http://127.0.0.1:8081/parse/classes/UserData/NgrgxOngKD", host: "dev-api.xxxxxx.xxxxx"
2017/11/04 12:07:01 [error] 2317#0: *116660 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 172.31.29.236, server: , request: "POST /parse/classes/UserData/KeoYxLNt1t HTTP/1.1", upstream: "http://127.0.0.1:8081/parse/classes/UserData/KeoYxLNt1t", host: "dev-api.xxxxx.xxxx"
2017/11/04 12:07:01 [error] 2317#0: *116657 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 172.31.12.25, server: , request: "POST /parse/classes/UserData/4Wyph78e9R HTTP/1.1", upstream: "http://127.0.0.1:8081/parse/classes/UserData/4Wyph78e9R", host: "dev-api.xxxx.xxxxx"
2017/11/04 17:41:01 [error] 2317#0: *132762 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 172.31.29.236, server: , request: "POST /parse/classes/UserData/4Wyph78e9R HTTP/1.1", upstream: "http://127.0.0.1:8081/parse/classes/UserData/4Wyph78e9R", host: "dev-api.xxxxx.xxxx"

**Comment by paulovitin** [comment]
The client.sendMail never will be reached, because you return response before that.

```js
Parse.Cloud.define('buy-giftcard', function (request, response) {

  // here stop the code
  return response.success("Hello!");

  client.sendEmail({
```

**Comment by montymxb** [comment]
@nharbo you should probably upgrade parse-server to something greater than **1.9.2**, or at least test upgrading against your current database. There have been numerous fixes since then, it's possible your issue may have been resolved in a subsequent patch in any of the versions since then.

**Comment by nharbo** [comment]
Thanks @paulovitin  - It was something I did for testing - but you're absolutely right! I just forgot to edit the code before pasting it here :-) But I found out, that the reason I was getting this error, was that the cloudcode function never returned a response - so I just moved the "return response.success()" down under the sns-method, and then everything worked perfectly.. :) 

And yes @montymxb - I should! :) I just haven't had any problems with the server as it is now, so thats why I haven't upgraded :) And now this works too, so I might just keep it like this for now :D But thanks for the heads up!
